Source/WebGPU/WGSL/wgslc.cpp

```diff
@@ -0,0 +1,144 @@
+{
+    fprintf(stderr, "Usage: wgsl [options] <file> <entrypoint>\n");
+    fprintf(stderr, "  -h|--help  Prints this help message\n");
+    fprintf(stderr, "  --dump-ast-after-checking  Dumps the AST after parsing and checking\n");
+    fprintf(stderr, "  --dump-ast-at-end  Dumps the AST after generating code\n");
+    fprintf(stderr, "  --dump-generated-code  Dumps the generated Metal code\n");
+    fprintf(stderr, "\n");
+
+    exit(help ? EXIT_SUCCESS : EXIT_FAILURE);
+}
+
+struct CommandLine {
+public:
+    CommandLine(int argc, char** argv)
+    {
+        parseArguments(argc, argv);
+    }
+
+    const char* file() const { return m_file; }
+    const char* entrypoint() const { return m_entrypoint; }
+    bool dumpASTAfterCheck() const { return m_dumpASTAfterCheck; }
+    bool dumpASTAtEnd() const { return m_dumpASTAtEnd; }
+    bool dumpGeneratedCode() const { return m_dumpGeneratedCode; }
+
+private:
+    void parseArguments(int, char**);
+
+    const char* m_file { nullptr };
+    const char* m_entrypoint { nullptr };
+    bool m_dumpASTAfterCheck { false };
+    bool m_dumpASTAtEnd { false };
+    bool m_dumpGeneratedCode { false };
+};
+
+void CommandLine::parseArguments(int argc, char** argv)
+{
+    for (int i = 1; i < argc; ++i) {
+        const char* arg = argv[i];
+        if (!strcmp(arg, "-h") || !strcmp(arg, "--help"))
+            printUsageStatement(true);
+
+        if (!strcmp(arg, "--dump-ast-after-checking")) {
+            m_dumpASTAfterCheck = true;
+            continue;
+        }
+
+        if (!strcmp(arg, "--dump-ast-at-end")) {
+            m_dumpASTAtEnd = true;
+            continue;
+        }
+
+        if (!strcmp(arg, "--dump-generated-code")) {
+            m_dumpGeneratedCode = true;
+            continue;
+        }
+
+        if (!m_file)
+            m_file = arg;
+        else if (!m_entrypoint)
+            m_entrypoint = arg;
+        else
+            printUsageStatement(false);
+    }
+
+    if (!m_file || !m_entrypoint)
+        printUsageStatement(false);
+}
+
+static int runWGSL(const CommandLine& options)
+{
+    WGSL::Configuration configuration;
+
+
+    String fileName = String::fromLatin1(options.file());
+    auto readResult = FileSystem::readEntireFile(fileName);
+    if (!readResult.has_value()) {
+        // FIXME: print error message
+        return EXIT_FAILURE;
+    }
+
+    auto source = String::fromUTF8WithLatin1Fallback(readResult->data(), readResult->size());
+    auto checkResult = WGSL::staticCheck(source, std::nullopt, configuration);
+    if (auto* failedCheck =  std::get_if<WGSL::FailedCheck>(&checkResult)) {
+        for (const auto& error : failedCheck->errors)
+            dataLogLn(error);
+        return EXIT_FAILURE;
+    }
+
+    auto& shaderModule = std::get<WGSL::SuccessfulCheck>(checkResult).ast;
+    if (options.dumpASTAfterCheck())
+        WGSL::AST::dumpAST(shaderModule);
+
+    String entrypointName = String::fromLatin1(options.entrypoint());
```

Comment:
Are you sure latin1 is right for this?

Reply:
I didn't anticipate passing any other kinds of strings in test programs, but maybe it's useful? Are you implying it should be utf8?

Reply:
yeah, utf8 seems better 🤔